chain/client/src/chunk_validation.rs

```diff
@@ -700,16 +700,38 @@ impl Client {
         }
 
         let chunk_header = chunk.cloned_header();
-        let chunk_validators = self
-            .epoch_manager
-            .get_chunk_validator_assignments(
-                epoch_id,
-                chunk_header.shard_id(),
-                chunk_header.height_created(),
-            )?
-            .ordered_chunk_validators();
+        let chunk_validators_assignment = self.epoch_manager.get_chunk_validator_assignments(
+            epoch_id,
+            chunk_header.shard_id(),
+            chunk_header.height_created(),
+        )?;
+        let mut chunk_validators = chunk_validators_assignment.ordered_chunk_validators();
         let witness =
             self.create_state_witness(prev_chunk_header, chunk, transactions_storage_proof)?;
+
+        if let Some(my_signer) = self.validator_signer.clone() {
+            if chunk_validators_assignment.contains(my_signer.validator_id()) {
+                // Since we've created the state witness, directly send the chunk endorsement to block producers.
+                // We can bypass the state witness validation logic.
+                send_chunk_endorsement_to_block_producers(
+                    &chunk_header,
+                    self.epoch_manager.as_ref(),
+                    my_signer.as_ref(),
+                    &self.chunk_validator.network_sender,
+                );
+
+                // remove ourselves from the list of chunk validators to send the chunk state witness to.
+                chunk_validators.retain(|account_id| account_id != my_signer.validator_id());
+
+                // Additionally send a copy of the chunk endorsement to ourselves.
+                // Mainly useful in tests where we don't have a good way to handle network messages and there's
```

Comment:
Ah... this makes me a bit uncomfortable :( If a test doesn't handle network messages causing the endorsements to not propagate, the correct thing to do should be to fix the test, not to change production behavior. A similar problem happened when the ShardsManager was refactored out, which broke a ton of tests because they weren't handling network messages, and the solution there was to make a SynchronousShardsManagerAdapter. It's not great, really, but if we can do something like that for the problem here, I'd prefer that to hacking the code here.

If we must hack the code here then may I suggest that we only send ourselves the endorsement, if we are a block producer?

Btw I'm also not super comfortable with just skipping the validation of our own witness. It doesn't save any latency (as we still have to wait for others to validate), and if our witness somehow turns out to be correct, we're going to be signing an invalid state witness just because we produced it. So that makes me think that this is purely a change to ease testing.

Reply:
Hmm, I see... In that case we can consider skipping over this optimization and for the short term measure of getting in changes for the MVP, send the endorsement to ourselves and add a major TODO for refactoring?

Reply:
Yeah, sounds reasonable. Thanks!

Reply:
https://github.com/near/nearcore/issues/10508